lf-queue: Rewrite based on Nikolaev's paper and hazard pointers

csm1735 opened the pull request:
Instead of using the old implementation, a concurrent queue 'lfq' is provided, which can utilize hazard pointers to release memory.

The program has been rewritten using C11 Atomics to support other processor architectures, and it has solved the data race issues, enabling it to pass the runtime detection of Thread Sanitizer.

jserv commented on lf-queue/Makefile:
```diff
@@ -1,14 +1,17 @@
-CFLAGS =  -Wall -O2 -g -I.
-LDFLAGS = -lpthread
+all: 
```
Check https://github.com/sysprog21/jitboy/blob/master/Makefile for modularized build rules.

csm1735 replied:
Done!

jserv commented on lf-queue/README.md:
```diff
@@ -1,8 +1,8 @@
-* multithreaded
-* bounded
-* lock-free
+The original code was from [quiz8_test3](https://hackmd.io/@sysprog/linux2023-quiz8#%E6%B8%AC%E9%A9%97-3) in the course "Linux Kernel Internals" at NCKU during Spring 2023.
```
Remove this paragraph since this repository is meant to be complementary.

jserv commented on lf-queue/README.md:
```diff
@@ -1,8 +1,8 @@
-* lock-free
+The original code was from [quiz8_test3](https://hackmd.io/@sysprog/linux2023-quiz8#%E6%B8%AC%E9%A9%97-3) in the course "Linux Kernel Internals" at NCKU during Spring 2023.
+
+However, the original code was specifically designed for x86(-64) architecture and had data race issues that could not pass the runtime detection of Thread Sanitizer.
```
Remove this paragraph.

csm1735 replied:
Done!

jserv commented on lf-queue/README.md:
```diff
@@ -1,8 +1,8 @@
+
+However, the original code was specifically designed for x86(-64) architecture and had data race issues that could not pass the runtime detection of Thread Sanitizer.
+
+Therefore, this program has been modified using C11 Atomics to support other processor architectures and has rewritten the code to solve the data race issues, enabling it to pass the runtime detection of Thread Sanitizer.
```
Move the description to the beginning of this document.

csm1735 replied:
Done!

jserv commented on lf-queue/lfq.c:
Is the assertion valid? If not, remove it.

csm1735 replied:
It is not valid. Removed.

jserv commented on lf-queue/lfq.c:
Check if the statement `TODO` is valid.

csm1735 replied:
It is not valid. Removed.

jserv commented on lf-queue/lfq.c:
It is not necessary to have `DEBUG` here. Simply write down `assert`.

jserv replied:
Quote from https://man7.org/linux/man-pages/man3/assert.3.html
> If the macro NDEBUG is defined at the moment <assert.h> was last included, the macro assert() generates no code, and hence does nothing at all.

csm1735 replied:
Done!

jserv commented on lf-queue/atomics.h:
```diff
@@ -0,0 +1,26 @@
+#define ATOMIC_ADD64 ATOMIC_ADD
+#define mb() atomic_thread_fence(memory_order_seq_cst)
+
+/* compiler barrier only: runtime reordering is impossible on x86 */
```
This comment is outdated. Refine.
Use Linux kernel style naming. See https://www.kernel.org/doc/Documentation/memory-barriers.txt 
i.e., `smp_mb`, `smp_wmb`, and `smp_rmb`.

csm1735 replied:
Done!

jserv commented on lf-queue/README.md:
```diff
@@ -1,8 +1,4 @@
-# Multithreaded Bounded Lock-free Queue
+# lfq
+This program has been modified using C11 Atomics to support other processor architectures and has rewritten the code to solve the data race issues, enabling it to pass the runtime detection of Thread Sanitizer.
```
Instead of mentioning "This program has been modified using C11 Atomics," you shall give a general overview for the rewritten program. i.e., some wording as following:
> The lock-free queue implemented is based on the paper [A Scalable, Portable, and Memory-Efficient
Lock-Free FIFO Queue](https://drops.dagstuhl.de/opus/volltexte/2019/11335/pdf/LIPIcs-DISC-2019-28.pdf) by Ruslan Nikolaev.
> Nikolaev's paper provides an overview of various methods for implementing lock-free queues. This implementation specifically adopts the SCQ (Scalable Circular Queue) approach, which is designed for bounded queues. However, this approach can be easily extended to support unbounded FIFO queues capable of storing an arbitrary number of elements.

You shall check Nikolaev's paper as well.

jserv commented on lf-queue/lfq.c:
```diff
@@ -0,0 +1,214 @@
+    struct lfq_node *old_tail = XCHG(&ctx->fpt, node); /* seq_cst */
+    atomic_store(&old_tail->free_next, node);
+}
+static void free_pool(struct lfq_ctx *ctx, bool freeall)
```
Separate each function with a blank line properly.

jserv commented on lf-queue/atomics.h:
```diff
@@ -0,0 +1,27 @@
+#define ATOMIC_RELEASE atomic_flag_clear
+
+#define ATOMIC_SUB atomic_fetch_sub
+#define ATOMIC_SUB64 ATOMIC_SUB
```
`ATOMIC_SUB64` is not really used. Drop it.

jserv commented on lf-queue/atomics.h:
```diff
@@ -0,0 +1,27 @@
+#define ATOMIC_SUB atomic_fetch_sub
+#define ATOMIC_SUB64 ATOMIC_SUB
+#define CAS atomic_compare_exchange_strong
+#define cmpxchg CAS
```
Ditto. Drop `cmpxchg`.

jserv commented on lf-queue/lfq.c:
Replace `!(old_tail->next)` with `!old_tail->next`.

jserv commented on lf-queue/atomics.h:
```diff
@@ -0,0 +1,27 @@
+#include <stdlib.h>
+#include <string.h>
+
+#define ATOMIC_SET atomic_flag_test_and_set
```
Check if both `ATOMIC_SET` and `ATOMIC_RELEASE` are used or not.

csm1735 replied:
Both ATOMIC_SET and ATOMIC_RELEASE are not used.

jserv replied:
Then, remove unused code. Always keep the code clean and elegant.

jserv commented on lf-queue/lfq.c:
```diff
@@ -0,0 +1,214 @@
+        struct lfq_node *p = ctx->fph;
+        if ((!atomic_load(&p->can_free)) || (!atomic_load(&p->free_next)) ||
+            in_hp(ctx, (struct lfq_node *) p))
+            goto final;
```
Instead of using `goto`, can you simply use `break` statement?

csm1735 replied:
Done!

jserv commented on lf-queue/lfq.c:
Validate this comment for the sake of C11 atomics transition.

csm1735 replied:
It is not valid. Removed.

jserv commented on lf-queue/README.md:
```diff
@@ -1,8 +1,5 @@
-# Multithreaded Bounded Lock-free Queue
+# lfq
+The lock-free queue implemented is based on the paper [A Scalable, Portable, and Memory-Efficient Lock-Free FIFO Queue](https://drops.dagstuhl.de/opus/volltexte/2019/11335/pdf/LIPIcs-DISC-2019-28.pdf) by Ruslan Nikolaev.
```
Replace "implemented" with "implementation."

jserv commented on lf-queue/README.md:
```diff
@@ -1,8 +1,5 @@
-* multithreaded
-* bounded
-* lock-free
+In addition, this concurrent queue utilizes Hazard pointers to release memory.
```
Proofread as following:
>  In addition, this implementation employs a hazard pointer-based memory reclamation system for concurrent queues.

jserv commented on lf-queue/Makefile:
```diff
@@ -1,14 +1,21 @@
-CFLAGS =  -Wall -O2 -g -I.
+CC = gcc
```
Remove this line since we may use clang as well. Let `make` decide the default compiler i.e., `CC`.

csm1735 committed:
lf-queue: Rewrite based on Nikolaev's paper and hazard pointers

Instead of using the old implementation, a concurrent queue based
on Nikolaev's paper and hazard pointers is provided.

The program has been rewritten using C11 Atomics to support other
processor architectures, and it has solved the data race issues,
enabling it to pass the runtime detection of Thread Sanitizer.

jserv commented:
Thank @csm1735 for contributing!
